### PR TITLE
Add docs for OPENVINO_FRAMEWORK_MAP macro

### PR DESCRIPTION
**Pull request by mateusztabaka**
Ticket: 98762

### CODE REVIEW

File: docs/Extensibility_UG/frontend_extensions.md [hunk 1]
```diff
@@ -88,6 +88,32 @@ So the conclusion is that each attribute of target OpenVINO operation should be
 
 This is achieved by specifying maps as arguments for `OpExtension` constructor.
 
+### Mapping custom operations to frontends with OPENVINO_FRAMEWORK_MAP macro
+
+> **NOTE**: Below solution works only for ONNX and Tensorflow frontends.
+
+`OPENVINO_FRAMEWORK_MAP` is a macro that should be used inside OpenVINO operation's class definition and that let you specify the mapping between this operation to a frontend operation.
+
+Let's consider following example. Imagine you have an ONNX model with `CustomOp` operation (and this operation has `mode` attribute) and a Tensorflow model with `CustomOpV3` operation (this operation has `axis` attribute) and both of them can be implemented with a single OpenVINO operation `CustomOp` like follows:
```
Comment (PiotrKrzem) [inline docs/Extensibility_UG/frontend_extensions.md:97]:
```suggestion
Let's consider the following example. Imagine you have an ONNX model with `CustomOp` operation (and this operation has `mode` attribute) and a Tensorflow model with `CustomOpV3` operation (this operation has `axis` attribute) and both of them can be implemented with a single OpenVINO operation `CustomOp` like follows:
```

File: docs/Extensibility_UG/frontend_extensions.md [hunk 2]
```diff
@@ -88,6 +88,32 @@ So the conclusion is that each attribute of target OpenVINO operation should be
 
 This is achieved by specifying maps as arguments for `OpExtension` constructor.
 
+### Mapping custom operations to frontends with OPENVINO_FRAMEWORK_MAP macro
+
+> **NOTE**: Below solution works only for ONNX and Tensorflow frontends.
+
+`OPENVINO_FRAMEWORK_MAP` is a macro that should be used inside OpenVINO operation's class definition and that let you specify the mapping between this operation to a frontend operation.
```
Comment (PiotrKrzem) [inline docs/Extensibility_UG/frontend_extensions.md:95]:
```suggestion
`OPENVINO_FRAMEWORK_MAP` is a macro that should be used inside OpenVINO operation's class definition and that lets you specify the mapping between this operation to a frontend operation.
```

File: docs/Extensibility_UG/frontend_extensions.md [hunk 3]
```diff
@@ -88,6 +88,32 @@ So the conclusion is that each attribute of target OpenVINO operation should be
...
+
+Let's consider following example. Imagine you have an ONNX model with `CustomOp` operation (and this operation has `mode` attribute) and a Tensorflow model with `CustomOpV3` operation (this operation has `axis` attribute) and both of them can be implemented with a single OpenVINO operation `CustomOp` like follows:
+
+@snippet ov_extensions.cpp frontend_extension_framework_map_macro_headers
+@snippet ov_extensions.cpp frontend_extension_framework_map_macro_CustomOp
+
+Let's take a closer look on the parameters this macro takes:
```
Comment (PiotrKrzem) [inline docs/Extensibility_UG/frontend_extensions.md:102]:
```suggestion
Let's take a closer look at the parameters this macro takes:
```

File: docs/Extensibility_UG/frontend_extensions.md [hunk 4]
```diff
@@ -88,6 +88,32 @@ So the conclusion is that each attribute of target OpenVINO operation should be
...
+```cpp
+OPENVINO_FRAMEWORK_MAP(framework, name, attributes_map, attributes_values)
+```
+- `framework` - framework name.
+- `name` - the framework operation name. It's optional if the OpenVINO custom operation name (that is the name that is passed as the first parameter to `OPENVINO_OP` macro) is the same as framework operation name and both `attributes_map` and `attributes_values` are not provided.
```
Comment (PiotrKrzem) [inline docs/Extensibility_UG/frontend_extensions.md:107]:
```suggestion
- `name` - the framework operation name. It's optional if the OpenVINO custom operation name (that is the name that is passed as the first parameter to `OPENVINO_OP` macro) is the same as the framework operation name and both `attributes_map` and `attributes_values` are not provided.
```

File: docs/Extensibility_UG/frontend_extensions.md [hunk 5]
```diff
@@ -88,6 +88,32 @@ So the conclusion is that each attribute of target OpenVINO operation should be
...
+- `attributes_map` - used to provide mapping between OpenVINO operation attribute and framework operation attribute. Contains key-value pairs, where key is an OpenVINO operation attribute name and value is its corresponding framework operation attribute name. This parameter is optional if number of OpenVINO operation attributes and their names match one-to-one with framework operation attributes.
```
Comment (PiotrKrzem) [inline docs/Extensibility_UG/frontend_extensions.md:108]:
```suggestion
- `attributes_map` - used to provide a mapping between OpenVINO operation attribute and framework operation attribute. Contains key-value pairs, where key is an OpenVINO operation attribute name and value is its corresponding framework operation attribute name. This parameter is optional if the number of OpenVINO operation attributes and their names match one-to-one with framework operation attributes.
```

File: docs/Extensibility_UG/frontend_extensions.md [hunk 6]
```diff
@@ -88,6 +88,32 @@ So the conclusion is that each attribute of target OpenVINO operation should be
...
+- `attributes_values` - used to provide default values for OpenVINO operation attributes that are not specified in `attributes_map`. Contains key-value pairs, where key is an OpenVINO operation attribute name and the value is this attribute value. This parameter cannot be provided if `attributes_map` contains all of OpenVINO operation attributes or if `attributes_map` is not provided.
+
+In the example above, `OPENVINO_FRAMEWORK_MAP` is used twice.
+First, OpenVINO `CustomOp` is mapped to ONNX `CustomOp` operation, `m_mode` attribute is mapped to `mode` attribute, while `m_axis` attribute gets default value `-1`.
```
Comment (PiotrKrzem) [inline docs/Extensibility_UG/frontend_extensions.md:112]:
```suggestion
First, OpenVINO `CustomOp` is mapped to ONNX `CustomOp` operation, `m_mode` attribute is mapped to `mode` attribute, while `m_axis` attribute gets the default value `-1`.
```

File: docs/Extensibility_UG/frontend_extensions.md [hunk 7]
```diff
@@ -88,6 +88,32 @@ So the conclusion is that each attribute of target OpenVINO operation should be
...
+Secondly, OpenVINO `CustomOp` is mapped to Tensorflow `CustomOpV3` operation, `m_axis` attribute is mapped to `axis` attribute, while `m_mode` attribute gets default value `"linear"`.
```
Comment (PiotrKrzem) [inline docs/Extensibility_UG/frontend_extensions.md:113]:
```suggestion
Secondly, OpenVINO `CustomOp` is mapped to Tensorflow `CustomOpV3` operation, `m_axis` attribute is mapped to `axis` attribute, while `m_mode` attribute gets the default value `"linear"`.
```

**Review by PiotrKrzem — COMMENTED** [review]
Some grammatical suggestions

**Review by itikhono — APPROVED** [review]

File: docs/snippets/ov_extensions.cpp [hunk 8]
```diff
@@ -40,6 +46,27 @@ class CustomOperation : public ov::op::Op {
...
 };
 
+//! [frontend_extension_framework_map_macro_CustomOp]
+class CustomOp : public ov::op::Op {
+    std::string m_mode;
+    int m_axis;
+
+public:
+    OPENVINO_OP("CustomOp");
+    OPENVINO_FRAMEWORK_MAP(onnx, "CustomOp", { {"m_mode", "mode"} }, { {"m_axis", -1} });
```
Comment (ilyachur) [inline docs/snippets/ov_extensions.cpp:56]: Why do you need to have new custom operation? Can you reuse operation from the `template_extension`?

Comment (mateusztabaka) [reply docs/snippets/ov_extensions.cpp:56]: I could reuse Identity op from template_extension, but I also wanted to show how to map attributes.

File: docs/snippets/ov_extensions.cpp [hunk 9]
```diff
@@ -125,5 +152,12 @@ ov::Core core;
 core.add_extension("openvino_template_extension.so");
 //! [add_extension_lib]
 }
+
+{
+//! [frontend_extension_framework_map_macro_add_extension]
+ov::Core core;
+core.add_extension(ov::frontend::OpExtension<CustomOp>());
```
Comment (slyalin) [inline docs/snippets/ov_extensions.cpp:159]: It was designed differently. It should be enough to register it as `ov::OpExtension<CustomOp>` which should register it for both FW conversion and IR reading purposes. So the idea is to register regular OV op which has FW-extensions inside, and those extensions will be registered automatically while registering `CustomOp` as an OV op.

Comment (slyalin) [reply docs/snippets/ov_extensions.cpp:159]: @itikhono, is that correct?

**Review by slyalin — CHANGES_REQUESTED** [review]

File: docs/Extensibility_UG/frontend_extensions.md [hunk 10]
```diff
@@ -88,6 +88,32 @@ So the conclusion is that each attribute of target OpenVINO operation should be
 
 This is achieved by specifying maps as arguments for `OpExtension` constructor.
 
+### Mapping custom operations to frontends with OPENVINO_FRAMEWORK_MAP macro
+
+> **NOTE**: Below solution works only for ONNX and Tensorflow frontends.
```
[the remark this answers is not in the document]
Comment (mateusztabaka) [reply docs/Extensibility_UG/frontend_extensions.md:93]: We do have a ticket, but it's not implemented yet.

File: docs/snippets/ov_extensions.cpp (hunk 9 above)
Comment (slyalin) [reply docs/snippets/ov_extensions.cpp:159]: !!!

Comment (slyalin) [reply docs/snippets/ov_extensions.cpp:159]: This incorrect code still exists in documentation!

**Pull request merged**